File: examples/python/ex15_callbot_streaming/main.py
"""15 Callbot Streaming — multi-level streaming pipeline (audio → VAD → STT
→ LLM router → TTS).

Tier 1 — pure compute, no API keys (everything is mocked). Run from this
directory:

    uv sync
    uv run python main.py
"""

from __future__ import annotations

import asyncio

from operonx.core import END, START, Operon, graph, op


@op
async def customer_audio(sample_count: int):
    """Mic input — yields fixed 32 ms audio chunks."""
    for i in range(sample_count):
        await asyncio.sleep(0.005)
        yield {"audio": f"chunk_{i}", "timestamp_ms": i * 32}


@op
async def vad(audio: str, timestamp_ms: int):
    """Voice Activity Detection — N-to-M generator.

    Variable-length speech segments out of fixed-size input. Silence
    chunks yield zero outputs.
    """
    speech_timestamps = {64, 128}
    if timestamp_ms in speech_timestamps:
        yield {
            "segment": f"speech_from_{audio}",
            "start_ms": timestamp_ms,
            "end_ms": timestamp_ms + 32,
        }


@op
def stt(segment: str, start_ms: int, end_ms: int):
    return {"transcript": f"Hello from {segment} [{start_ms}-{end_ms}ms]"}


@op
def classify_intent(transcript: str):
    if "hello" in transcript.lower():
        return {"intent": "greeting", "confidence": 0.95}
    return {"intent": "general", "confidence": 0.8}


@op
def handle_intent(intent: str, transcript: str):
    if intent == "greeting":
        return {"response": "Hello! How can I help you today?"}
    return {"response": f"I understand. Let me help with: {transcript}"}


@graph
def llm_router(transcript):
    """Nested @graph — classify_intent → handle_intent."""
    c = classify_intent(transcript=transcript)
    h = handle_intent(intent=c["intent"], transcript=transcript)
    START >> c >> h >> END


@op
async def tts(response: str):
    """TTS — yields audio chunks word by word."""
    for i, word in enumerate(response.split()):
        await asyncio.sleep(0.003)
        yield {"audio_out": f"tts_{i}_{word}", "index": i}


@graph
def callbot(samples):
    """audio → vad → stt → llm_router → tts."""
    audio = customer_audio(sample_count=samples)
    v = vad(audio=audio["audio"], timestamp_ms=audio["timestamp_ms"])
    transcribe = stt(
        segment=v["segment"],
        start_ms=v["start_ms"],
        end_ms=v["end_ms"],
    )
    router = llm_router(transcript=transcribe["transcript"])
    speak = tts(response=router["response"])
    START >> audio >> v >> transcribe >> router >> speak >> END


async def main() -> None:
    g = callbot(samples=5)
    result = await Operon(g).run(inputs={})
    content = {k: v for k, v in result.items() if k != "$state"}
    print(f"[callbot] {content}")


if __name__ == "__main__":
    asyncio.run(main())


# ── the served front door ───────────────────────────────────────────────
# The same pipeline behind a real socket. `[[serve]]` in operonx.toml
# boots it with `session = "per_connection"`: one websocket, one
# long-lived run, and the run always finishes — the peer disconnecting
# ends `ingress`, the graph drains, and whatever must happen after the
# caller is gone still happens.
#
# `customer_audio` is replaced by the connection itself: each websocket
# message is one audio frame, "payload@timestamp_ms". TTS chunks go back
# out the same socket through `egress`.
from operonx.core.serve import egress, ingress


@op
def unpack_frame(item=None) -> dict:
    """One websocket message → the audio chunk and when it was captured."""
    audio, _, ts = str(item).partition("@")
    return {"audio": audio, "timestamp_ms": int(ts or 0)}


@graph
def served_callbot():
    frames = ingress()
    unpack = unpack_frame(item=frames["item"])
    v = vad(audio=unpack["audio"], timestamp_ms=unpack["timestamp_ms"])
    transcribe = stt(
        segment=v["segment"],
        start_ms=v["start_ms"],
        end_ms=v["end_ms"],
    )
    router = llm_router(transcript=transcribe["transcript"])
    speak = tts(response=router["response"])
    out = egress(item=speak["audio_out"])
    START >> frames >> unpack >> v >> transcribe >> router >> speak >> out >> END

File: examples/rust/ex15_callbot_streaming/src/main.rs
//! 15 Callbot Streaming — Rust-side demo.
//!
//! Mirrors `examples/python/ex15_callbot_streaming/main.py`.
//!
//! ⚠️  Rust-runtime limitation: the callbot is built almost entirely on
//! generator ops + nested `@graph` composition — neither of which the
//! Rust scheduler exposes yet. The ops are declared here so the Python
//! side can serialise its `graph.json` and the Rust binary can be
//! exercised structurally; per-item dispatch and the nested router will
//! short-circuit until those land.

use operonx::{op, Operon};
use serde_json::Value;

#[op(name = "customer_audio")]
fn customer_audio(sample_count: i64) -> Value {
    let out: Vec<Value> = (0..sample_count)
        .map(|i| serde_json::json!({ "audio": format!("chunk_{i}"), "timestamp_ms": i * 32 }))
        .collect();
    serde_json::json!({ "items": out })
}

#[op(name = "vad")]
fn vad(audio: String, timestamp_ms: i64) -> Value {
    let speech_timestamps = [64_i64, 128_i64];
    if speech_timestamps.contains(&timestamp_ms) {
        serde_json::json!({
            "segment": format!("speech_from_{audio}"),
            "start_ms": timestamp_ms,
            "end_ms": timestamp_ms + 32,
        })
    } else {
        serde_json::json!({
            "segment": Value::Null,
            "start_ms": Value::Null,
            "end_ms": Value::Null,
        })
    }
}

#[op(name = "stt")]
fn stt(segment: String, start_ms: i64, end_ms: i64) -> Value {
    serde_json::json!({
        "transcript": format!("Hello from {segment} [{start_ms}-{end_ms}ms]"),
    })
}

#[op(name = "classify_intent")]
fn classify_intent(transcript: String) -> Value {
    if transcript.to_lowercase().contains("hello") {
        serde_json::json!({ "intent": "greeting", "confidence": 0.95 })
    } else {
        serde_json::json!({ "intent": "general", "confidence": 0.8 })
    }
}

#[op(name = "handle_intent")]
fn handle_intent(intent: String, transcript: String) -> Value {
    if intent == "greeting" {
        serde_json::json!({ "response": "Hello! How can I help you today?" })
    } else {
        serde_json::json!({ "response": format!("I understand. Let me help with: {transcript}") })
    }
}

#[op(name = "tts")]
fn tts(response: String) -> Value {
    let out: Vec<Value> = response
        .split_whitespace()
        .enumerate()
        .map(|(i, word)| serde_json::json!({ "audio_out": format!("tts_{i}_{word}"), "index": i }))
        .collect();
    serde_json::json!({ "items": out })
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let here = std::path::Path::new(env!("CARGO_MANIFEST_DIR"));
    let graph_bundle: Value =
        serde_json::from_slice(&std::fs::read(here.join("graph.json"))?)?;
    let inputs_bundle: Value =
        serde_json::from_slice(&std::fs::read(here.join("inputs.json"))?)?;

    let graph_v = graph_bundle
        .get("callbot")
        .ok_or("graph.json missing `callbot` entry")?;
    let graph_json = serde_json::to_string(graph_v)?;

    let inputs_obj = inputs_bundle
        .get("callbot")
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();

    let engine = Operon::builder(&graph_json).auto_register().build()?;
    match engine.run_json(inputs_obj, None, None, None) {
        Ok(r) => println!("[callbot] {r}"),
        Err(e) => println!("[callbot] error: {e}"),
    }

    Ok(())
}
